In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import time

In [2]:
y2015 = pd.read_csv('LoanStats3d.csv',skipinitialspace=True, header=1)

c:\users\shuaix\appdata\local\programs\python\python37\lib\site-packages\IPython\core\interactiveshell.py:3020: DtypeWarning: Columns (0,19,55) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
y2015 = y2015[:-2]
y2015.dropna(axis=1, thresh=20000,inplace=True)

In [4]:
y2015['int_rate'] = pd.to_numeric(y2015['int_rate'].str.strip('%'), errors='coerce')

y2015.drop(['emp_title', 'zip_code', 'earliest_cr_line', 'revol_util',
            'sub_grade', 'addr_state'], 1, inplace=True)

In [5]:
pd.get_dummies(y2015)

loan_amnt  funded_amnt  funded_amnt_inv  int_rate  installment  \
0         35000.0      35000.0          35000.0     14.85       829.90   
1         10400.0      10400.0          10400.0     22.45       289.91   
2         20000.0      20000.0          20000.0     10.78       432.66   
3          3600.0       3600.0           3600.0     13.99       123.03   
4         24700.0      24700.0          24700.0     11.99       820.28   
5         11950.0      11950.0          11950.0     13.44       405.18   
6         16000.0      16000.0          16000.0     12.88       363.07   
7         20000.0      20000.0          20000.0      9.17       637.58   
8          1400.0       1400.0           1400.0     12.88        47.10   
9         18000.0      18000.0          18000.0     19.48       471.70   
10        22400.0      22400.0          22400.0     12.88       508.30   
11        10000.0      10000.0          10000.0      6.49       306.45   
12         8000.0       8000.0           8000.0     11.48       263.74   
13        20000.0      20000.0          20000.0      8.49       631.26   
14        28000.0      28000.0          28000.0      6.49       858.05   
15        23000.0      23000.0          23000.0      8.49       471.77   
16         8650.0       8650.0           8650.0     19.89       320.99   
17        29900.0      29900.0          29900.0     12.88       678.49   
18         9600.0       9600.0           9600.0      7.49       298.58   
19        16000.0      16000.0          16000.0     14.85       379.39   
20        18000.0      18000.0          18000.0     11.99       400.31   
21        25000.0      25000.0          25000.0      7.49       777.55   
22        16000.0      16000.0          16000.0     12.88       363.07   
23         4200.0       4200.0           4200.0     13.99       143.53   
24        12000.0      12000.0          12000.0      9.17       382.55   
25         4000.0       4000.0           4000.0      9.17       127.52   
26        14000.0      14000.0          14000.0      9.80       450.43   
27        10500.0      10500.0          10500.0     15.77       367.96   
28        10000.0      10000.0          10000.0     11.48       329.67   
29        15000.0      15000.0          15000.0      9.80       482.61   
...           ...          ...              ...       ...          ...   
421065     4200.0       4200.0           4200.0     15.99       147.64   
421066    20000.0      20000.0          20000.0      6.99       617.46   
421067    15000.0      15000.0          15000.0     11.99       498.15   
421068    13000.0      13000.0          13000.0      8.19       408.52   
421069    26950.0      26950.0          26950.0      8.19       846.88   
421070    15000.0      15000.0          15000.0     11.44       329.44   
421071     6000.0       6000.0           6000.0     17.86       216.50   
421072    18700.0      18700.0          18700.0     17.86       473.44   
421073    23000.0      23000.0          23000.0     15.59       554.32   
421074    25000.0      25000.0          25000.0     15.99       607.82   
421075    25000.0      25000.0          25000.0     15.59       602.52   
421076    26500.0      26500.0          26500.0      8.67       545.87   
421077     8000.0       8000.0           8000.0      8.19       251.40   
421078    21000.0      21000.0          21000.0      6.03       639.15   
421079    12000.0      12000.0          12000.0      9.49       384.34   
421080    10775.0      10775.0          10775.0      6.03       327.95   
421081     7000.0       7000.0           7000.0     14.31       240.30   
421082     6225.0       6225.0           6225.0     16.49       220.37   
421083    13150.0      13150.0          13150.0     14.99       312.77   
421084    10000.0      10000.0          10000.0      6.49       306.45   
421085     4000.0       4000.0           4000.0      8.67       126.59   
421086    10850.0      10850.0          10850.0     19.24       399.04   
421087     7

In [6]:
from sklearn import ensemble
from sklearn.model_selection import cross_val_score

start_time = time.time()
rfc = ensemble.RandomForestClassifier()
X = y2015.drop('loan_status', 1)
Y = y2015['loan_status']
X = pd.get_dummies(X)
X = X.dropna(axis=1)

cross_val_score(rfc, X, Y, cv=10)

c:\users\shuaix\appdata\local\programs\python\python37\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
c:\users\shuaix\appdata\local\programs\python\python37\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
c:\users\shuaix\appdata\local\programs\python\python37\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
c:\users\shuaix\appdata\local\programs\python\python37\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in ve

array([0.99045448, 0.99069193, 0.99000285, 0.99237693, 0.992258  ,
       0.9915693 , 0.99237675, 0.99301795, 0.99225782, 0.99323153])

In [7]:
print("--- %s seconds ---" % (time.time() - start_time))

--- 183.68471813201904 seconds ---


In [9]:
from sklearn import tree
start_time = time.time()
decision_tree = tree.DecisionTreeClassifier(
    criterion='entropy',
    max_features=1,
    max_depth=4  
)

cross_val_score(decision_tree, X, Y, cv=10)
print("--- %s seconds ---" % (time.time() - start_time))

--- 29.804687976837158 seconds ---


## random forest with no PCA took 184 seconds while decision tree took 30 seconds.